sky/shell/platform/android/io/flutter/view/FlutterView.java

```diff
@@ -472,13 +473,50 @@ public void runFromBundle(String bundlePath, String snapshotPath) {
         postRun();
     }
 
-    public void runFromFile(String main, String packageRoot) {
-        preRun();
-        mSkyEngine.runFromFile(main, packageRoot, "");
-        postRun();
+    private class RunFromSourceRunnable implements Runnable {
+        RunFromSourceRunnable(String main,
+                              String packages,
+                              String assetsDirectory,
+                              FlutterView flutterView) {
+            this.main = main;
+            this.packages = packages;
+            this.assetsDirectory = assetsDirectory;
+            this.flutterView = flutterView;
+        }
+
+        private String main;
+        private String packages;
+        private String assetsDirectory;
+        private FlutterView flutterView;
+
+        public void run() {
+            preRun();
+            flutterView.mSkyEngine.runFromFile(main, packages, assetsDirectory);
+            postRun();
+            synchronized (this) {
+                notify();
+            }
+        }
+    }
+
+    public void runFromSource(String main,
+                              String packages,
+                              String assetsDirectory) {
+        RunFromSourceRunnable runnable =
+            new RunFromSourceRunnable(main, packages, assetsDirectory, this);
+        post(runnable);
+        try {
+            synchronized (runnable) {
+                runnable.wait();
```

Comment:
There is a potential race here if the task calls notify() before this thread enters the wait().  Move the post() call into the synchronized block, or consider android.os.ConditionVariable or java.util.concurrent.Semaphore